File: baekjoon/2504/Main.java
import java.util.*;
import java.io.*;

public class Main {

	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringBuilder sb = new StringBuilder();

		String str = br.readLine();

		Stack<Character> st = new Stack<>();
		int ans = 0;
		int temp = 1;

		for(int i=0; i<str.length(); i++){

			if(str.charAt(i)=='('){
				st.push('(');
				temp = temp * 2;
			}else if(str.charAt(i)=='['){
				st.push('[');
				temp = temp * 3;
			}else if(str.charAt(i)==')'){
				if(st.empty()||st.peek()!='('){
					ans = 0;
					break;
				}else if(str.charAt(i-1)=='('){
					ans = ans + temp;
				}		
					st.pop();
					temp = temp / 2;
			}else if(str.charAt(i)==']'){
				if(st.empty()||st.peek()!='['){
					ans =0;
					break;
				}else if(str.charAt(i-1)=='['){
					ans = ans + temp;
				}
					st.pop();
					temp = temp / 3;
			}
		}

		if(st.empty()){
			System.out.println(ans);
		}else{
			System.out.println(0);
		}


	}
}

File: baekjoon/2504/Main.py



str = list(input())

st = []
ans =0
temp =1

for i in range(len(str)):

    if str[i] == '(':
        st.append('(')
        temp = temp * 2

    elif str[i] == '[':
        st.append('[')
        temp = temp * 3

    elif str[i] == ')':
        if not st or st[-1] != '(':
            ans = 0
            break
        elif str[i-1] == '(':
            ans = ans + temp
        st.pop()
        temp = temp / 2

    elif str[i] ==']':
        if not st or st[-1] != '[':
            ans = 0
            break
        elif str[i-1] == '[':
            ans = ans + temp
        del st[-1]
        temp = temp / 3

if(st):
    print(0)
else:
    print(int(ans))